libressl-build/share/man/man3/dsa.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "dsa 3"
.TH dsa 3 "2017-01-09" "LibreSSL " "LibreSSL"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dsa \- Digital Signature Algorithm
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\& #include <openssl/dsa.h>
\& #include <openssl/engine.h>
\&
\& DSA *  DSA_new(void);
\& void   DSA_free(DSA *dsa);
\&
\& int    DSA_size(const DSA *dsa);
\&
\& DSA *  DSA_generate_parameters(int bits, unsigned char *seed,
\&                int seed_len, int *counter_ret, unsigned long *h_ret,
\&                void (*callback)(int, int, void *), void *cb_arg);
\&
\& DH *   DSA_dup_DH(const DSA *r);
\&
\& int    DSA_generate_key(DSA *dsa);
\&
\& int    DSA_sign(int dummy, const unsigned char *dgst, int len,
\&                unsigned char *sigret, unsigned int *siglen, DSA *dsa);
\& int    DSA_sign_setup(DSA *dsa, BN_CTX *ctx, BIGNUM **kinvp,
\&                BIGNUM **rp);
\& int    DSA_verify(int dummy, const unsigned char *dgst, int len,
\&                const unsigned char *sigbuf, int siglen, DSA *dsa);
\&
\& void DSA_set_default_method(const DSA_METHOD *meth);
\& const DSA_METHOD *DSA_get_default_method(void);
\& int DSA_set_method(DSA *dsa, const DSA_METHOD *meth);
\& DSA *DSA_new_method(ENGINE *engine);
\& const DSA_METHOD *DSA_OpenSSL(void);
\&
\& int DSA_get_ex_new_index(long argl, char *argp, int (*new_func)(),
\&             int (*dup_func)(), void (*free_func)());
\& int DSA_set_ex_data(DSA *d, int idx, char *arg);
\& char *DSA_get_ex_data(DSA *d, int idx);
\&
\& DSA_SIG *DSA_SIG_new(void);
\& void   DSA_SIG_free(DSA_SIG *a);
\& int    i2d_DSA_SIG(const DSA_SIG *a, unsigned char **pp);
\& DSA_SIG *d2i_DSA_SIG(DSA_SIG **v, unsigned char **pp, long length);
\&
\& DSA_SIG *DSA_do_sign(const unsigned char *dgst, int dlen, DSA *dsa);
\& int    DSA_do_verify(const unsigned char *dgst, int dgst_len,
\&             DSA_SIG *sig, DSA *dsa);
\&
\& DSA *  d2i_DSAPublicKey(DSA **a, unsigned char **pp, long length);
\& DSA *  d2i_DSAPrivateKey(DSA **a, unsigned char **pp, long length);
\& DSA *  d2i_DSAparams(DSA **a, unsigned char **pp, long length);
\& int    i2d_DSAPublicKey(const DSA *a, unsigned char **pp);
\& int    i2d_DSAPrivateKey(const DSA *a, unsigned char **pp);
\& int    i2d_DSAparams(const DSA *a,unsigned char **pp);
\&
\& int    DSAparams_print(BIO *bp, const DSA *x);
\& int    DSAparams_print_fp(FILE *fp, const DSA *x);
\& int    DSA_print(BIO *bp, const DSA *x, int off);
\& int    DSA_print_fp(FILE *bp, const DSA *x, int off);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
These functions implement the Digital Signature Algorithm (\s-1DSA\s0).  The
generation of shared \s-1DSA\s0 parameters is described in
\&\fIDSA_generate_parameters\fR\|(3);
\&\fIDSA_generate_key\fR\|(3) describes how to
generate a signature key. Signature generation and verification are
described in \fIDSA_sign\fR\|(3).
.PP
The \fB\s-1DSA\s0\fR structure consists of several \s-1BIGNUM\s0 components.
.PP
.Vb 10
\& struct
\&        {
\&        BIGNUM *p;              // prime number (public)
\&        BIGNUM *q;              // 160\-bit subprime, q | p\-1 (public)
\&        BIGNUM *g;              // generator of subgroup (public)
\&        BIGNUM *priv_key;       // private key x
\&        BIGNUM *pub_key;        // public key y = g^x
\&        // ...
\&        }
\& DSA;
.Ve
.PP
In public keys, \fBpriv_key\fR is \s-1NULL.\s0
.PP
Note that \s-1DSA\s0 keys may use non-standard \fB\s-1DSA_METHOD\s0\fR implementations,
either directly or by the use of \fB\s-1ENGINE\s0\fR modules. In some cases (eg. an
\&\s-1ENGINE\s0 providing support for hardware-embedded keys), these \s-1BIGNUM\s0 values
will not be used by the implementation or may be used for alternative data
storage. For this reason, applications should generally avoid using \s-1DSA\s0
structure elements directly and instead use \s-1API\s0 functions to query or
modify keys.
.SH "CONFORMING TO"
.IX Header "CONFORMING TO"
\&\s-1US\s0 Federal Information Processing Standard \s-1FIPS 186 \s0(Digital Signature
Standard, \s-1DSS\s0), \s-1ANSI X9.30\s0
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIbn\fR\|(3), \fIdh\fR\|(3), \fIerr\fR\|(3), \fIrand\fR\|(3),
\&\fIrsa\fR\|(3), \fIsha\fR\|(3), \fIengine\fR\|(3),
\&\fIDSA_new\fR\|(3),
\&\fIDSA_size\fR\|(3),
\&\fIDSA_generate_parameters\fR\|(3),
\&\fIDSA_dup_DH\fR\|(3),
\&\fIDSA_generate_key\fR\|(3),
\&\fIDSA_sign\fR\|(3), \fIDSA_set_method\fR\|(3),
\&\fIDSA_get_ex_new_index\fR\|(3),
\&\fIRSA_print\fR\|(3)


Changes
Revision history for H2O.

2.3.0-beta1 TBD
	- [access-log] log connections that closed prior to sending a response #1235 (Kazuho Oku)
	- [ssl] support redis as a data store for session cache / ticket #1087 (Ichito Nagata)
	- [misc] include process id in crash backtrace #1254 (Kazuho Oku)

2.2.2 2017-04-23 03:24:00+0000
	- [ssl] fix OCSP stapling error when LibreSSL is used #1275 (Ian Moone)


2.2.1 2017-04-22 22:43:00+0000
	- [mruby] correct the line number reported on an exception #1239 #1251 (Ichito Nagata)
	- [mruby] retain the order of request headers sharing a single name #1271 (Kazuho Oku)
	- [ssl] fix assertion failure in `decode_ssl_input` #1264 (Kazuho Oku)
	- [ssl] fix OCSP stapling error when OpenSSL 1.1.0 is used #1270 (Kazuho Oku)
	- [libh2o] fix crash when abruptly closing an HTTP/2 connection on libuv #1250 (Kazuho Oku)
	- [libh2o] fix memory leak of `_timestamp_cache` #1255 (Kazuho Oku)
	- [doc] restore doc of `%{...}e` #1252 (Kazuho Oku)
	- [doc] fix typo suggesting using `brotli` instead of `br` #1263 (Bogdan Khomutsky)
	- [misc] fix undefined behaviors detected by ubsan #1246 (Frederik Deweerdt)

2.2.0 2017-04-05 02:18:00+0000
	- [core] add `crash-handler.wait-pipe-close` parameter #1092 (Frederik Deweerdt)
	- [core] introduce an option to bypass the `server` header sent from upstream #1226 (Frederik Deweerdt)
	- [core] apply global- and host-level configuration to requests not applicable to any of the path-level configurations #1231 (Kazuho Oku)
	- [access-log] add `%{remote}p` for logging the remote port #1166 (Kazuho Oku)
	- [access-log] add support for JSON-style escapes and null #1208 (Kazuho Oku)
	- [access-log] add specifier for logging per-request environment variables #1221 (Yannick Koechlin)
	- [access-log] add support for `<`, `>` modifiers for logging either the original or the final response #1238 (Kazuho Oku)
	- [access-log] do not emit request-total-time twice #1017 (Kazuho Oku)
	- [fastcgi] fix a bug that closes the FastCGI listener socket during startup #1203 (Frederik Deweerdt)
	- [file] add directive for serving gzipped files, decompressing them on-the-fly #1140 (Ichito Nagata)
	- [headers] fix buffer overrun during startup #1180 (Frederik Deweerdt)
	- [http1][proxy] preserve the cases of characters used in header names #1194 (Frederik Deweerdt)
	- [http1][proxy] fix undefined behavior in HTTP/1 parser #1189 (Frederik Deweerdt)
	- [http1] stop reading from socket after sending 400 to avoid the risk of assertion failure #1223 (Frederik Deweerdt)
	- [http2] recognize `x-http2-push-only` attribute on `link` header #1169 (Frederik Deweerdt)
	- [http2] add optional timeout for closing connections upon graceful shutdown #1108 (Frederik Deweerdt)
	- [http2] do not ack an acked PING frame #1175 (Moto Ishisawa)
	- [http2] reject requests exceeding the maximum allowed size more efficiently #1183 (Frederik Deweerdt)
	- [mruby] remove dependenty to mkmf #1197 (Yuki Kurihara)
	- [mruby] correct the line number reported on an exception #1239 (Ichito Nagata)
	- [proxy] add directives for tweaking headers sent to upstream #1126 (Justin Zhu)
	- [proxy] retain case-sensitivity of unix socket paths #1131 (Frederik Deweerdt)
	- [proxy] add directive for controlling the `via` request header #1225 (Frederik Deweerdt)
	- [ssl] add directive for logging session ID #1164 (Yannick Koechlin)
	- [ssl] add support for TLS 1.3 draft-18 #1204 (Kazuho Oku)
	- [ssl] stop evicting session entries in memcached when they are removed from internal cache #1185 (Ichito Nagata)
	- [ssl] fix crash when a secp384r1, secp521r1 certificate is used with TLS 1.3 #1214 (Kazuho Oku)
	- [ssl] fix build failure with OpenSSL 1.1.0 #1216 (Kazuho Oku)
	- [ssl] add doc for `handshake-timeout` #1233 (Kazuho Oku)
	- [status] fix race condition during start-up #1242 (Frederik Deweerdt)
	- [libh2o] implement `h2o_evloop_destroy` #1200 (kazan417)
	- [misc] add test code for fuzzing #1174 #1182 #1191 #1192 (Frederik Deweerdt, Jonathan Foote)
	- [misc] fix issues reported by Coverity #1168 #1172 #1179 (Harrison Bowden, Frederik Deweerdt)

2.1.0 2017-01-17 23:43:27+0000
	- [core] TCP latency optimization #873 #1076 (Kazuho Oku)
	- [core] provide tag to include other YAML files from the configuration file #1022 (Ichito Nagata)
	- [core] accept sequence of mappings for path-level configuration #1042 (Ichito Nagata)
	- [core] fix broken support for TCP Fast Open in OS X #1065 (Ichito Nagata)
	- [access-log] provide directive to emit request-level errors #1075 (Kazuho Oku)
	- [access-log] emit values of all `set-cookie` headers concatenated #1161 (Kazuho Oku)
	- [fastcgi] fix connection failure when `fastcgi.spawn` is used with an uid #1119 (Kazuho Oku)
	- [file] more pre-defined MIME types #1103 (Joe Duarte)
	- [http2][proxy] recognize link rel=preload headers in interim response as a trigger to push resources #916 (Kazuho Oku)
	- [http1][http2] validate characters used in the headers #974 #1044 (Frederik Deweerdt, Kazuho Oku)
	- [http1][http2] notify error downstream when an error occurred while generating a response #1031 (Frederik Deweerdt)
	- [http1][http2] fix resource leak upon upgrade failure to HTTP/2 #1161 (Frederik Deweerdt)
	- [http2] add `http2-push-preload` directive to turn off H2 push being initiated by link rel=preload header #929 (Kazuho Oku)
	- [http2] add support for `cache-digest` header #967 #988 (Kazuho Oku)
	- [http2] drop `host` header in HTTP/2 layer #973 #998 (Frederik Deweerdt, Kazuho Oku)
	- [http2] don't use etag for calculating casper cookie #986 (Kazuho Oku)
	- [http2] add support for H2 debug state #1019 (Ichito Nagata)
	- [mruby] add dos_detector mruby handler #1013 (Ichito Nagata)
	- [mruby] add DSL for access control lists (acl) #1016 (Ichito Nagata)
	- [mruby] share mruby state and constants between handlers #1032 (Ichito Nagata)
	- [mruby] add library for address-block-based access control #1038 (Ichito Nagata)

libressl-2.4.5/apps/openssl/dsa.c
/* $OpenBSD: dsa.c,v 1.7 2015/10/17 07:51:10 semarie Exp $ */
/* Copyright (C) 1995-1998 Eric Young (eay@cryptsoft.com)
 * All rights reserved.
 *
 * This package is an SSL implementation written
 * by Eric Young (eay@cryptsoft.com).
 * The implementation was written so as to conform with Netscapes SSL.
 *
 * This library is free for commercial and non-commercial use as long as
 * the following conditions are aheared to.  The following conditions
 * apply to all code found in this distribution, be it the RC4, RSA,
 * lhash, DES, etc., code; not just the SSL code.  The SSL documentation
 * included with this distribution is covered by the same copyright terms
 * except that the holder is Tim Hudson (tjh@cryptsoft.com).
 *
 * Copyright remains Eric Young's, and as such any Copyright notices in
 * the code are not to be removed.
 * If this package is used in a product, Eric Young should be given attribution
 * as the author of the parts of the library used.
 * This can be in the form of a textual message at program startup or
 * in documentation (online or textual) provided with the package.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *    "This product includes cryptographic software written by
 *     Eric Young (eay@cryptsoft.com)"
 *    The word 'cryptographic' can be left out if the rouines from the library
 *    being used are not cryptographic related :-).
 * 4. If you include any Windows specific code (or a derivative thereof) from
 *    the apps directory (application code) you must include an acknowledgement:
 *    "This product includes software written by Tim Hudson (tjh@cryptsoft.com)"
 *
 * THIS SOFTWARE IS PROVIDED BY ERIC YOUNG ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * The licence and distribution terms for any publically available version or
 * derivative of this code cannot be changed.  i.e. this code cannot simply be
 * copied and put under another distribution licence
 * [including the GNU Public Licence.]
 */

#include <openssl/opensslconf.h>	/* for OPENSSL_NO_DSA */

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

#include "apps.h"

#include <openssl/bio.h>
#include <openssl/bn.h>
#include <openssl/dsa.h>
#include <openssl/err.h>
#include <openssl/evp.h>
#include <openssl/pem.h>
#include <openssl/x509.h>

static struct {
	const EVP_CIPHER *enc;
	char *infile;
	int informat;
	int modulus;
	int noout;
	char *outfile;
	int outformat;
	char *passargin;
	char *passargout;
	int pubin;
	int pubout;
	int pvk_encr;
	int text;
} dsa_config;

static int
dsa_opt_enc(int argc, char **argv, int *argsused)
{
	char *name = argv[0];

	if (*name++ != '-')
		return (1);

	if ((dsa_config.enc = EVP_get_cipherbyname(name)) != NULL) {
		*argsused = 1;
		return (0);
	}

	return (1);
}

static struct option dsa_options[] = {
	{
		.name = "in",
		.argname = "file",
		.desc = "Input file (default stdin)",
		.type = OPTION_ARG,
		.opt.arg = &dsa_config.infile,
	},
	{
		.name = "inform",
		.argname = "format",
		.desc = "Input format (PEM (default) or any other supported"
		    " format)",
		.type = OPTION_ARG_FORMAT,
		.opt.value = &dsa_config.informat,
	},
	{
		.name = "noout",
		.desc = "No output",
		.type = OPTION_FLAG,
		.opt.flag = &dsa_config.noout,
	},
	{
		.name = "out",
		.argname = "file",
		.desc = "Output file (default stdout)",
		.type = OPTION_ARG,
		.opt.arg = &dsa_config.outfile,
	},
	{
		.name = "outform",
		.argname = "format",
		.desc = "Output format (DER, MSBLOB, PEM (default) or PVK)",
		.type = OPTION_ARG_FORMAT,
		.opt.value = &dsa_config.outformat,
	},
	{
		.name = "passin",
		.argname = "source",
		.desc = "Input file passphrase source",
		.type = OPTION_ARG,
		.opt.arg = &dsa_config.passargin,
	},
	{
		.name = "passout",
		.argname = "source",
		.desc = "Output file passphrase source",
		.type = OPTION_ARG,
		.opt.arg = &dsa_config.passargout,
	},
	{
		.name = "pubin",
		.desc = "Read a public key from the input file instead of"
		    " private key",
		.type = OPTION_FLAG,
		.opt.flag = &dsa_config.pubin,
	},
	{
		.name = "pubout",
		.desc = "Output a public key instead of private key",
		.type = OPTION_FLAG,
		.opt.flag = &dsa_config.pubout,
	},
	{
		.name = "pvk-none",
		.desc = "PVK encryption level",
		.type = OPTION_VALUE,
		.value = 0,
		.opt.value = &dsa_config.pvk_encr,
	},
	{
		.name = "pvk-strong",
		.desc = "PVK encryption level (default)",
		.type = OPTION_VALUE,
		.value = 2,
		.opt.value = &dsa_config.pvk_encr,
	},
	{
		.name = "pvk-weak",
		.desc = "PVK encryption level",
		.type = OPTION_VALUE,
		.value = 1,
		.opt.value = &dsa_config.pvk_encr,
	},
	{
		.name = "text",
		.desc = "Print the key in text form",
		.type = OPTION_FLAG,
		.opt.flag = &dsa_config.text,
	},
	{
		.name = NULL,
		.type = OPTION_ARGV_FUNC,
		.opt.argvfunc = dsa_opt_enc,
	},
	{ NULL },
};

static void
show_ciphers(const OBJ_NAME *name, void *arg)
{
	static int n;

	if (!islower((unsigned char)*name->name))
		return;

	fprintf(stderr, " -%-24s%s", name->name, (++n % 3 ? "" : "\n"));
}

static void
dsa_usage(void)
{
	fprintf(stderr,
	    "usage: dsa [-in file] [-inform format] [-noout]\n"
	    "    [-out file] [-outform format] [-passin src] [-passout src]\n"
	    "    [-pubin] [-pubout] [-pvk-none | -pvk-strong | -pvk-weak]\n"
	    "    [-text] [-ciphername]\n\n");
	options_usage(dsa_options);
	fprintf(stderr, "\n");

	fprintf(stderr, "Valid ciphername values:\n\n");
	OBJ_NAME_do_all_sorted(OBJ_NAME_TYPE_CIPHER_METH, show_ciphers, NULL);
	fprintf(stderr, "\n");
}

int
dsa_main(int argc, char **argv)
{
	int ret = 1;
	DSA *dsa = NULL;
	int i;
	BIO *in = NULL, *out = NULL;
	char *passin = NULL, *passout = NULL;

	if (single_execution) {
		if (pledge("stdio rpath wpath cpath tty", NULL) == -1) {
			perror("pledge");
			exit(1);
		}
	}

	memset(&dsa_config, 0, sizeof(dsa_config));

	dsa_config.pvk_encr = 2;
	dsa_config.informat = FORMAT_PEM;
	dsa_config.outformat = FORMAT_PEM;

	if (options_parse(argc, argv, dsa_options, NULL, NULL) != 0) {
		dsa_usage();
		goto end;
	}

	if (!app_passwd(bio_err, dsa_config.passargin, dsa_config.passargout,
	    &passin, &passout)) {
		BIO_printf(bio_err, "Error getting passwords\n");
		goto end;
	}

	in = BIO_new(BIO_s_file());
	out = BIO_new(BIO_s_file());
	if (in == NULL || out == NULL) {
		ERR_print_errors(bio_err);
		goto end;
	}
	if (dsa_config.infile == NULL)
		BIO_set_fp(in, stdin, BIO_NOCLOSE);
	else {
		if (BIO_read_filename(in, dsa_config.infile) <= 0) {
			perror(dsa_config.infile);
			goto end;
		}
	}

	BIO_printf(bio_err, "read DSA key\n");

	{
		EVP_PKEY *pkey;

		if (dsa_config.pubin)
			pkey = load_pubkey(bio_err, dsa_config.infile,
			    dsa_config.informat, 1, passin, "Public Key");
		else
			pkey = load_key(bio_err, dsa_config.infile,
			    dsa_config.informat, 1, passin, "Private Key");

		if (pkey) {
			dsa = EVP_PKEY_get1_DSA(pkey);
			EVP_PKEY_free(pkey);
		}
	}
	if (dsa == NULL) {
		BIO_printf(bio_err, "unable to load Key\n");
		ERR_print_errors(bio_err);
		goto end;
	}
	if (dsa_config.outfile == NULL) {
		BIO_set_fp(out, stdout, BIO_NOCLOSE);
	} else {
		if (BIO_write_filename(out, dsa_config.outfile) <= 0) {
			perror(dsa_config.outfile);
			goto end;
		}
	}

	if (dsa_config.text) {
		if (!DSA_print(out, dsa, 0)) {
			perror(dsa_config.outfile);
			ERR_print_errors(bio_err);
			goto end;
		}
	}
	if (dsa_config.modulus) {
		fprintf(stdout, "Public Key=");
		BN_print(out, dsa->pub_key);
		fprintf(stdout, "\n");
	}
	if (dsa_config.noout)
		goto end;
	BIO_printf(bio_err, "writing DSA key\n");
	if (dsa_config.outformat == FORMAT_ASN1) {
		if (dsa_config.pubin || dsa_config.pubout)
			i = i2d_DSA_PUBKEY_bio(out, dsa);
		else
			i = i2d_DSAPrivateKey_bio(out, dsa);
	} else if (dsa_config.outformat == FORMAT_PEM) {
		if (dsa_config.pubin || dsa_config.pubout)
			i = PEM_write_bio_DSA_PUBKEY(out, dsa);
		else
			i = PEM_write_bio_DSAPrivateKey(out, dsa, dsa_config.enc,
			    NULL, 0, NULL, passout);
#if !defined(OPENSSL_NO_RSA) && !defined(OPENSSL_NO_RC4)
	} else if (dsa_config.outformat == FORMAT_MSBLOB ||
	    dsa_config.outformat == FORMAT_PVK) {
		EVP_PKEY *pk;
		pk = EVP_PKEY_new();
		EVP_PKEY_set1_DSA(pk, dsa);
		if (dsa_config.outformat == FORMAT_PVK)
			i = i2b_PVK_bio(out, pk, dsa_config.pvk_encr, 0,
			    passout);
		else if (dsa_config.pubin || dsa_config.pubout)
			i = i2b_PublicKey_bio(out, pk);
		else
			i = i2b_PrivateKey_bio(out, pk);
		EVP_PKEY_free(pk);
#endif
	} else {
		BIO_printf(bio_err, "bad output format specified for outfile\n");
		goto end;
	}
	if (i <= 0) {
		BIO_printf(bio_err, "unable to write private key\n");
		ERR_print_errors(bio_err);
	} else
		ret = 0;
end:
	BIO_free(in);
	if (out != NULL)
		BIO_free_all(out);
	if (dsa != NULL)
		DSA_free(dsa);
	free(passin);
	free(passout);

	return (ret);
}
